src/util/ecdsa.rs

```diff
@@ -413,6 +415,61 @@ impl<'de> ::serde::Deserialize<'de> for PublicKey {
     }
 }
 
+/// An ECDSA signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct EcdsaSig {
+    /// The underlying ECDSA Signature
+    pub sig: secp256k1::Signature,
+    /// The corresponding hash type
+    pub hash_ty: EcdsaSigHashType,
+}
+
+impl EcdsaSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, EcdsaSigError> {
+        let (hash_ty, sig) = sl.split_last()
+            .ok_or(EcdsaSigError::InvalidEcdsaSig)?;
+        let hash_ty = EcdsaSigHashType::from_u32_standard(*hash_ty as u32)
+            .map_err(|_| EcdsaSigError::NonStandardSigHashType(*hash_ty))?;
+        let sig = secp256k1::Signature::from_der(sig)
+            .map_err(|_| EcdsaSigError::InvalidEcdsaSig)?;
+        Ok(EcdsaSig { sig, hash_ty })
+    }
+
+    /// Serialize EcdsaSig
+    pub fn to_vec(&self) -> Vec<u8> {
+        // TODO: add support to serialize to a writer to SerializedSig
+        let mut ser_sig = self.sig.serialize_der().to_vec();
+        ser_sig.push(self.hash_ty.as_u32() as u8);
+        ser_sig
+    }
+}
+
+/// A key-related error.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum EcdsaSigError {
+    /// Base58 encoding error
+    NonStandardSigHashType(u8),
+    /// secp256k1-related error
+    InvalidEcdsaSig,
+}
+
+
+impl fmt::Display for EcdsaSigError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            EcdsaSigError::NonStandardSigHashType(hash_ty) =>
+                write!(f, "Non standard signature hash type {}", hash_ty),
+            EcdsaSigError::InvalidEcdsaSig => write!(f, "Invalid Ecdsa Sig"),
```

Comment:
Since this message is shown to the user it could be a bit nicer. E.g. "Invalid ECDSA signature"

Can be done in a separate PR because it doesn't affect API.